### ISSUE TITLE
Troubles commanding regulators and capacitors with 8500 node system

### ISSUE THREAD

**Issue by blthayer** [GRIDAPPSD/GOSS-GridAPPS-D#755]
### Description of Issue
- commanding regulators causes simulation to crash
- capacitor commands go through, but may or may not actually change capacitor position

### Affected Version
develop branch

### Steps to Reproduce
1. Checkout the `pyvvo` branch of the `gridappsd-docker` repository
2. Checkout the `develop` branch of the `gridappsd-pyvvo` repository
  - Note: This repo is a mess. @blthayer is working on a cleaner and more well tested version in the `pyvvo` repo
3. In terminal, navigate to `gridappsd-docker` repository
4. Ensure old containers, dumps, etc. are removed and everything is up to date.
  - Note: There's a new shell script, `clean_dumps.sh` that helps with cleanup
5. Start docker-compose by executing `./run.sh`
6. Once inside the gridappsd container, execute `./run-gridappsd.sh` to start the platform
7. Open a new terminal, navigate to `gridappsd-docker` repo, and execute `docker-compose logs -f pyvvo` to get the pyvvo logs
8. In your browser, navigate to `localhost:8080`
   1.  Open the `Simulations` pane
   2. If the `Application Configuration` heading does not show up, you'll need to hit `Cancel`, refresh, and open `Simulations` again.
   3. In the `Line name` drop-down, select `ieee8500`
   4. Change the `Duration` to something higher, like `480`
   5. In the `Application name` drop-down, select `pyvvo`
   6. Click the `Submit` button
   7. After the feeder network graph has loaded, click `Start simulation`
9. Pay attention to the pyvvo logs in the second terminal you opened. While the genetic algorithm is running, you'll see something like the following at the bottom of the log:
   ```
   pyvvo_1        | [15:05:09] [population:ga:331] [INFO]: All individuals put in modeling queue.
   ```
   After a while, new messages will indicate that the first generation has run, and we're running the second generation. There are only two generations. Wait a little while longer for generation two to finish.
10. The pyvvo logs will inform you about the "best individual." After that, it will print the various commands to be sent to the simulator. This includes:
   1. Four capacitor switch closures.
      - NOTE: The viz doesn't show these capacitors, and the voltage strip charts don't change at all, so it's difficult to tell if the commands did anything. I'll file a separate issue here: the viz app is only displaying the uncontrolled capacitors, not the controlled capacitors.
   2. Regulator commands for all four voltage regulators
      - NOTE: Not all regulators receive a command for all three of their phases
11. After pyvvo has sent regulator and capacitor commands, examine the platform logs.
   1. It appears capacitor commands are going through, as we'll see a set of messages similar to the following: ```Sending the following message to the simulator. {"1349918726": {"cap_capbank0a": {"switchA": "CLOSED"}}}```
   2. It seems that commands sent to `reg_FEEDER_REG`, `reg_VREG2`, and `reg_VREG3` go through. However, it seems that commands to `reg_VREG4` crash the simulation. Immediately after commanding `reg_VREG4` we get an error like the following, which causes the simulation to crash.
      ```
      Thread-59 INFO [gov.pnnl.goss.gridappsd.simulation.SimulationProcess] - Simulator: ERROR    [14.447153d] : connection: local variable 'reg_VREG4.tapA' cannot be resolved
      ```
      - NOTE: The viz application doesn't seem to update the regulator taps which did change. This may be because the simulation crashes before outputting the change.

@afisher1 @poorva1209

**Comment by afisher1** [comment]
hey @blthayer. I can't build the pyvvo branch of gridappsd-docker. I'm getting docker compose file errors. first the version was bad and now I get this error. 

ERROR: The Compose file './docker-compose.yml' is invalid because:
services.mysql-pyvvo.volumes contains an invalid type, it should be a string.

It would be helpful is you could just provide me with the tap settings that you are trying to send to reg_VREG4 so I can investigate the crash. My guess is the settings you are sending cause GridLAB-D's powerflow to not converge.

**Comment by blthayer** [comment]
@afisher1 - I can try to get you the specific tap settings, but it'll be tomorrow until I can get to it.

As far as failing to build gridappsd-docker, it looks like I forgot to mention a few pre-requisite steps:

1. Make sure you have a modern version of docker-compose. I bumped the docker-compose file version
2. Checkout the 'pyvvo' branch of the gridappsd-python repository, and build the docker image, using the tag gridappsd/app-container-base
3. In the 'gridappsd-pyvvo' repository, build the docker image using the tag gridappsd-pyvvo

The build steps take a good while, so if you'd rather wait for me to get you the commands, I'll do so tomorrow.

**Comment by afisher1** [comment]
I think I'll wait till you can get me the comands as that would be easier to inject from the sample app.

**Comment by blthayer** [comment]
@afisher1 - here are all the commands my app sent, in order. Sorry for the delay. I had to copy+paste from a log, so you'll need to use something like notepad++ (hold down alt while dragging the cursor down to highlight a column) to strip out some of the junk.
```
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank0a".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:309] [INFO]: Capacitor command for cap_capbank0a:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:34.646507+00:00",
pyvvo_1        |             "difference_mrid": "7663801c-bb3e-4399-9e06-4c8b322bd301",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_A5866105-A527-F682-C982-69807C0E088B",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 0
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_A5866105-A527-F682-C982-69807C0E088B",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 1
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:311] [INFO]: Capacitor command for cap_capbank0a sent.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank0b".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:309] [INFO]: Capacitor command for cap_capbank0b:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:34.647181+00:00",
pyvvo_1        |             "difference_mrid": "0d7ab06d-ff71-4ac2-b993-a9cf71cdef83",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_614B787E-649F-6FFE-EA99-9260674DA020",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 0
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_614B787E-649F-6FFE-EA99-9260674DA020",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 1
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:311] [INFO]: Capacitor command for cap_capbank0b sent.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank0c".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:313] [INFO]: No command necessary for capacitor cap_capbank0c
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank1a".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:309] [INFO]: Capacitor command for cap_capbank1a:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:34.647856+00:00",
pyvvo_1        |             "difference_mrid": "806b0415-30a6-4b9c-93e9-c7427c02e027",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_A1395901-50BC-8B13-8CD6-F01B03CC8F65",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 0
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_A1395901-50BC-8B13-8CD6-F01B03CC8F65",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 1
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:311] [INFO]: Capacitor command for cap_capbank1a sent.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank1b".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:313] [INFO]: No command necessary for capacitor cap_capbank1b
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank1c".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:309] [INFO]: Capacitor command for cap_capbank1c:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:34.648485+00:00",
pyvvo_1        |             "difference_mrid": "c90c5da1-9397-4f98-8a23-2db6bb4c64dd",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_D188697C-E7E2-2433-B115-B5F5066C157A",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 0
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_D188697C-E7E2-2433-B115-B5F5066C157A",
pyvvo_1        |                     "attribute": "ShuntCompensator.sections",
pyvvo_1        |                     "value": 1
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:311] [INFO]: Capacitor command for cap_capbank1c sent.
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank2a".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:313] [INFO]: No command necessary for capacitor cap_capbank2a
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank2b".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:313] [INFO]: No command necessary for capacitor cap_capbank2b
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:287] [INFO]: Building command for capacitor "cap_capbank2c".
pyvvo_1        | [13:49:34] [pyvvo:command_capacitors:313] [INFO]: No command necessary for capacitor cap_capbank2c
pyvvo_1        | [13:49:34] [pyvvo:main:449] [WARNING]: Sleeping 5 seconds before commanding regulators.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:245] [INFO]: Building command for voltage regulator "reg_FEEDER_REG".
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:275] [INFO]: Regulator command for reg_FEEDER_REG:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:39.651830+00:00",
pyvvo_1        |             "difference_mrid": "863d4a2f-c046-4803-a82d-22659e2e5d48",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_330E7EDE-2C70-8F72-B183-AA4BA3C5E221",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 2
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_BBB3984D-2A67-7E15-0763-635C5B06A348",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 1
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_330E7EDE-2C70-8F72-B183-AA4BA3C5E221",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 3
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_BBB3984D-2A67-7E15-0763-635C5B06A348",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 0
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:277] [INFO]: Regulator command for reg_FEEDER_REG sent.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:245] [INFO]: Building command for voltage regulator "reg_VREG2".
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:275] [INFO]: Regulator command for reg_VREG2:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:39.652428+00:00",
pyvvo_1        |             "difference_mrid": "add10bff-1d51-4535-b93d-fd0f542a3e11",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_D4865B8A-86C4-F2CC-9C47-D7414F9AC721",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 9
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_98812C6A-8C1A-6B61-4601-8C6C00F342A2",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 6
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_D4865B8A-86C4-F2CC-9C47-D7414F9AC721",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 10
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_98812C6A-8C1A-6B61-4601-8C6C00F342A2",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 7
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:277] [INFO]: Regulator command for reg_VREG2 sent.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:245] [INFO]: Building command for voltage regulator "reg_VREG3".
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:275] [INFO]: Regulator command for reg_VREG3:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:39.653025+00:00",
pyvvo_1        |             "difference_mrid": "794e7853-1292-4ed2-af46-5db58901d3f5",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_E82651EB-3EA9-BB2A-DA35-9427F871713E",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 16
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_16BD9630-2D7C-7F26-EED6-7B4EF6454762",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 10
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_D8DDB71C-13CC-A730-635A-61E0EDBEE950",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 1
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_E82651EB-3EA9-BB2A-DA35-9427F871713E",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 13
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_16BD9630-2D7C-7F26-EED6-7B4EF6454762",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 9
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_D8DDB71C-13CC-A730-635A-61E0EDBEE950",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 2
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:277] [INFO]: Regulator command for reg_VREG3 sent.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:245] [INFO]: Building command for voltage regulator "reg_VREG4".
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:275] [INFO]: Regulator command for reg_VREG4:
pyvvo_1        | {
pyvvo_1        |     "command": "update",
pyvvo_1        |     "input": {
pyvvo_1        |         "simulation_id": "1245129973",
pyvvo_1        |         "message": {
pyvvo_1        |             "timestamp": "2018-11-02 13:49:39.653578+00:00",
pyvvo_1        |             "difference_mrid": "ae8fc174-0885-4f9b-8917-ed6822845a02",
pyvvo_1        |             "reverse_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_C847217F-0051-8B1C-A479-D98F1FE5AE43",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 11
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_B24BCA96-769F-1C25-673B-3D337BC26FC9",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 11
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_52E4D86E-9F6F-ED7B-7755-558415322D22",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 4
pyvvo_1        |                 }
pyvvo_1        |             ],
pyvvo_1        |             "forward_differences": [
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_C847217F-0051-8B1C-A479-D98F1FE5AE43",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 10
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_B24BCA96-769F-1C25-673B-3D337BC26FC9",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 10
pyvvo_1        |                 },
pyvvo_1        |                 {
pyvvo_1        |                     "object": "_52E4D86E-9F6F-ED7B-7755-558415322D22",
pyvvo_1        |                     "attribute": "TapChanger.step",
pyvvo_1        |                     "value": 7
pyvvo_1        |                 }
pyvvo_1        |             ]
pyvvo_1        |         }
pyvvo_1        |     }
pyvvo_1        | }.
pyvvo_1        | [13:49:39] [pyvvo:command_regulators:277] [INFO]: Regulator command for reg_VREG4 sent.
pyvvo_1        | [13:49:39] [pyvvo:<module>:542] [INFO]: Terminating pyvvo...

```

**Comment by afisher1** [comment]
The simulation was crashing because fncs_goss_bridge.py was converting all values to strings before publishing to fncs. This causes a type mismatch for all non string values expected on the GridLAB-D side of things. I showed @poorvasharma the fix and she made a pull request to the develop branch or GOSS-GridAPPS-D. I've confirmed that GridLAB-D no longer crashes when these messages are sent. 

@blthayer I want to caution you about sending multiple commands in a single timestep they should all be bundled together in one single message or GridLAB-D will only grab the late message recieved during that timestep. This is a known issue I need to fix on my end.
